atomicapp/nulecule/config.py

```diff
@@ -0,0 +1,123 @@
+import copy
+import logging
+
+from atomicapp.constants import (GLOBAL_CONF,
+                                 LOGGER_COCKPIT,
+                                 DEFAULT_PROVIDER,
+                                 DEFAULT_ANSWERS)
+from collections import defaultdict
+
+cockpit_logger = logging.getLogger(LOGGER_COCKPIT)
+
+
+class Config(object):
+    """
+    This class allows to store config data in different scopes along with
+    source info for the data. When fetching the value for a key in a scope,
+    the source info and the PRIORITY order of sources is taken into account.
+    """
+
+    PRIORITY = (
+        'cli',
+        'runtime',
+        'answers',
+        'defaults'
+    )
+
+    def __init__(self, answers=None, cli=None):
+        """
+        Initialize a Config instance.
+
+        Args:
+            answers (dict): Answers data
+            cli (dict): CLI data
+        """
+        answers = answers or {}
+        cli = cli or {}
+        self._data = defaultdict(defaultdict)
+        self._data['defaults'] = defaultdict(defaultdict)
+        self._data['answers'] = defaultdict(defaultdict)
+        self._data['cli'] = defaultdict(defaultdict)
+        self._data['runtime'] = defaultdict(defaultdict)
+
+        for scope, data in DEFAULT_ANSWERS.items():
+            for key, value in data.items():
+                self.set(key, value, scope=scope, source='defaults')
+        self.set('provider', DEFAULT_PROVIDER, scope=GLOBAL_CONF, source='defaults')
+
+        for scope, data in answers.items():
+            for key, value in data.items():
+                self.set(key, value, scope=scope, source='answers')
+
+        for key, value in cli.items():
+            self.set(key, value, scope=GLOBAL_CONF, source='cli')
+
+    def get(self, key, scope=GLOBAL_CONF):
+        """
+        Get the value of a key in a scope. This takes care of resolving
+        the value by going through the PRIORITY order of the various
+        sources of data.
+
+        Args:
+            key (str): Key
+            scope (str): Scope from which to fetch the value for the key
+
+        Returns:
+            Value for the key.
+        """
+        value = None
+        for source in self.PRIORITY:
+            value = self._data[source][scope].get(key)
+            if value:
+                break
+        return value
+
+    def set(self, key, value, scope=GLOBAL_CONF, source='defaults'):
```

Comment:
I would say we shouldn't have a default value for the source argument. We should force the caller to set that value IMO.